### PR TITLE
Add LGTM code quality badges

### PR DESCRIPTION
**Pull request by xcorail**
Hi there!

I thought you might be interested in adding these [code quality badges](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/ci/#badges) to your project. They will indicate a very high code quality to potential users and contributors.
To get an idea of the analyses reflected by these grades, check the alerts discovered by [LGTM](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown).

N.B.: I am on the team behind LGTM.com, I'd appreciate your feedback on this initiative, whether you're interested or not, if you find time to drop me a line. Thanks.

### CODE REVIEW

**Comment by xcorail** [comment]
Hey @AustinXII 
Thanks for the feedback. 

Can you please point me to the warnings you suspect to be false positives? 

If you find it cool, our recommended usage is intergation with Pull request, so that LGTM reports any new problem (or fix) directly within the Pull Request. It's a free [one-click](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/ci/) operation

**Comment by AustinXII** [comment]
An example would be it warning for ``if (target && target.whatever)`` because it expects target to always be true. Which isn’t wrong but we have the code in place just to prevent issues for other users on their servers

**Comment by xcorail** [comment]
I see. 
Having false positives or results that don't matter for your particular context is certainly the biggest pain point of code analysis tools. They will slow you down instead of helping you ...
So in your case, if this code is there for a good reason, that only you can know of, you can disable the alerts by adding a `//lgtm` comment in your code as described in our [documentation](https://lgtm.com/help/lgtm/alert-suppression). Keep only the alerts that matter to you.

**Comment by Zarel** [comment]
@xcorail Please ignore Austin, he doesn't speak for us.

@AustinXII I'm very confused how useless code is supposed to "prevent issues for other users on their servers". :|

**Comment by Zarel** [comment]
@xcorail I think this is really cool!

Some false positives include:

image

Which is code designed to shut TypeScript up. I don't actually know how to define a property to TypeScript without changing its existing value.

It looks like my options are to suppress it or report it to you guys. I think you guys might benefit from ignoring this error specifically when it's of the form `this.foo = this.foo` with a type doc comment above it saying `/** @type {foo} */` – that's an indication that it's a TypeScript property documentation, rather than completely useless.

**Comment by AustinXII** [comment]
Yeah I didn’t read all the errors at the time and made a bad false assumption on what would bring an alert considering the amount of errors, mb. I thought it would also be giving errors for lines like ``if (scavsRoom and ...)`` because of some of the other lines that were causing errors (eg. ``If (room and ...`` but those are actual bugs. @Zarel the scavsRoom assumption also addresses your complaint but tbf I did use a bad example to try and explain the issue that isn’t an issue. Also of course I don’t speak on behalf of PS I was sharing MY opinion on his project and tried to share a possible error to help get it integrated, I never once said my opinion was for PS, anyone can comment and there are labels to designate official project collaborators and owners :|

**Comment by Zarel** [comment]
I found another really interesting false positive here:

https://discuss.lgtm.com/t/false-positive-this-expression-always-evaluates-to-true/1270/1

**Comment by Zarel** [comment]
I'm not sure I like that comments are posted by my own account, like in: https://github.com/Zarel/Pokemon-Showdown/pull/4845#issuecomment-418972201

TravisBuddy uses its own GitHub account, which I think makes things much clearer.

**Comment by xcorail** [comment]
Hi @Zarel 

You are absolutely right.

On BitBucket, our comments are posted by LGTM, not by your personal account.
It will soon be the same on GitHub. We are currently working on migrating LGTM to a GitHub Apps API, and when it's done (soon), comments will be posted by LGTM

**Comment by esben-semmle** [comment]
Semmle JavaScript engnineer here. Thank you for the feedback!

> Which is code designed to shut TypeScript up. I don't actually know how to define a property to TypeScript without changing its existing value...

A fix for this is about to be merged, it should be online in about two weeks time (https://github.com/Semmle/ql/pull/166).

> I don't suppose you could support eslint's suppression for this specific problem?

We already support some suppression comments from linters, I will add this to the list of cases to consider. I do not have an estimate for when it will be online.

**Comment by xcorail** [comment]
Hey there

Just wanted to let you know that during 6 weeks, for each LGTM alert fixed, we'll make a donation to WWF. You can also win a free ticket to GitHub Universe, with travel and accommodation. Details here: https://competitions.lgtm.com/ghu-2018

**Comment by Zarel** [comment]
@xcorail @esben-semmle

My friend is having trouble suppressing these alerts:

https://github.com/Zarel/Pokemon-Showdown/pull/4922

**Comment by Zarel** [comment]
Okay, I just fixed all the LGTM issues across the three PS repositories.

There are 11 recommendations left which I can't fix:

- 7 unused variables in `globals.ts`, a file that defines globals for TypeScript.
  - The use-case is kind of weird, so I suggest either ignoring unused variables defined in non-module files (the file doesn't export anything)
  - Alternatively, allow reclassifying the file as a test from the web UI; I don't really want to create an `lgtm.yml` file just for LGTM to know to ignore the file
- 4 unused variables in `chat-commands.js` and `battle-stream.js`. These set up variables for use by `eval`.
  - I suggest ignoring unused vars in the same block as an `eval`.
  - Alternatively, respect `eslint-disable no-unused-vars`

**Comment by esben-semmle** [comment]
> Okay, I just fixed all the LGTM issues across the three PS repositories.

Excellent. 

> There are 11 recommendations left which I can't fix...

Thank you for the feedback.
We will have to consider the `globals.ts` situation a bit more, but I expect the `eval` issue to be fixed shortly.

**Comment by Zarel** [comment]
@esben-semmle Thanks! No need to deal with the `globals.ts` situation – apparently it was incorrectly exporting the globals anyway, and exporting them correctly (as in #4940) makes the LGTM errors go away.

**Comment by Zarel** [comment]
image

(from https://github.com/Zarel/Pokemon-Showdown/pull/4962)

This is concerning. This project uses `@param` to annotate types for TypeScript. In many cases, the parameter's variable name is sufficient documentation for what the parameter does, so the type is all that's needed.

It looks like `js/jsdoc/malformed-param-tag` is off by default, and so only appears in pull requests. I don't want to mislead contributors into thinking my project expects additional parameter documentation. Is there something I can do?

**Comment by esben-semmle** [comment]
Thank you for raising that issue, we will look into it.

**Comment by DubbleClick** [comment]
Unless I'm mistaken all lgtm warnings should be fixed with #5139

**Comment by esben-semmle** [comment]
Well done! 
Perhaps it is time to add a [badge](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/ci/#badges) you can be proud of? :)
[Language grade: JavaScript](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/context:javascript) 

I agree that the eight remaining alerts are not something that you need to fix, but I can not really see a satisfactory way to avoid flagging those locations without just whitelisting the locations manually.

**Comment by Zarel** [comment]
@esben-semmle, is there a way to whitelist a location on another line? I've been trying to enforce line length limits. Something similar to `// eslint-disable-next-line`

**Comment by esben-semmle** [comment]
> is there a way to whitelist a location on another line

No, not at the moment. Sorry.

File: README.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ Navigation: [Website][1] | **Server repository** | [Client repository][2] | [Dex
 [![Build Status](https://api.travis-ci.com/Zarel/Pokemon-Showdown.svg)](https://travis-ci.com/Zarel/Pokemon-Showdown)
 [![Dependency Status](https://david-dm.org/zarel/Pokemon-Showdown.svg)](https://david-dm.org/zarel/Pokemon-Showdown)
 [![devDependency Status](https://david-dm.org/zarel/Pokemon-Showdown/dev-status.svg)](https://david-dm.org/zarel/Pokemon-Showdown?type=dev)
+[![Code Quality: Javascript](https://img.shields.io/lgtm/grade/javascript/g/Zarel/Pokemon-Showdown.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/context:javascript)
```
Comment (Zarel) [inline README.md]:
```suggestion
```

Okay, I'll take the alert count, but the code quality badge seems redundant/vanity.

**Comment by Zarel** [comment]
https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/snapshot/80fe70d401acb3e562fcdbfcdeb9cbbf386119c4/files/src/battle-tooltips.ts?sort=name&dir=ASC&mode=heatmap#x73126d66521336f7:1

I got the error:

> This write to property 'comment' is useless, since another property write always overrides it.

But there's no write there; it's purely a declaration.

**Comment by esben-semmle** [comment]
Thanks, we are aware of that problem. A fix should be live in a weeks time. In the meantime, I have manually disabled that query for your project.

**Comment by Zarel** [comment]
Seemingly related bug:

https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-c1f391939b8a933ed0608f0ab4a1b1b3da7983f3

(Is your fix live yet? Does anything need to be done for that other query to be manually re-enabled?)

**Comment by esben-semmle** [comment]
> Is your fix live yet? 

The fix is live, and I have silently enabled the query for your project again.

> Seemingly related bug:

I am inclined to agree. But the initial values of those variables are indeed never read, and they could be ommitted.

Our treatmentment of destructuring assignments is not as good as for ordinary assignments (that is: we would not have flagged the assignments if they were on the form: `let targetLoc = 0, willMega = '', ...`).

**Comment by Zarel** [comment]
> Thanks, we are aware of that problem. A fix should be live in a weeks time. In the meantime, I have manually disabled that query for your project.

https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-0f9a0e9926364b4e5ae570cd10428bd92603195e

I don't know if this was overlooked or if it was a regression, but it looks like it's back.

**Pull request merged**